processing/src/main/java/org/apache/druid/query/dimension/DimensionSpec.java

```diff
@@ -70,6 +71,11 @@ default MultiValueDimensionVectorSelector decorate(MultiValueDimensionVectorSele
     throw new UOE("DimensionSpec[%s] cannot vectorize", getClass().getName());
   }
 
+  default VectorObjectSelector decorate(VectorObjectSelector selector)
```

Comment:
This is used with the poorly named `VectorColumnSelectorFactory.makeStringObjectSelector`, which should be renamed `VectorColumnSelectorFactory.makeDimensionObjectSelector` or something, which is for doing dimensiony things with a vector object selector similar to what is done for `SingleValueDimensionVectorSelector` or `MultiValueDimensionVectorSelector`. There are currently no implementations of this, like is the case with the other decorate methods with vector selectors, i was just trying for symmetry here so that the object selector could behave similar to the other string selectors when used in this context

Reply:
I think this makes sense, so if we're going to keep this method, that'd all be good stuff to add to the javadoc. This method does need one because it's not obvious from the signature how it works, what its preconditions are, and what kinds of objects will come out of the selector it returns.

Reply:
Will the VectorObjectSelector passed to this method always return an array whose values are all Strings? And is the same thing expected of the returned VectorObjectSelector? If so, please add that to the javadoc. If not, then I must be missing something and could you please explain it to me 🙂

By the way, what happens if the input is multi-value? Will some of the entries in the `Object[]` of either the input or the output VectorObjectSelector actually be `String[]` instead of `String`? Or is multi-value not supported today?

Reply:
I haven't yet implemented support for multi-value string object selectors, but your question also seems applicable if we were to add array types too. I guess the contents of the array would be dependent on whatever the underlying selector is providing, and it would be the responsibility of the `DimensionSpec` implementation to handle decoration appropriately? I guess we should also probably consider adding methods for the primitive numeric columns as well, since they are not masqueraded as "dimension selectors" in the vectorized engine, (if we ever want to fully support all of the non-vectorized behavior here I think).

I don't know, the more I think about this, do we really need `DimensionSpec`, like at all? The stuff it does is all over the place, and it feels like maybe this type of stuff should just be pushed down into either expression virtual columns or something more specialized as appropriate (more generic predicate filtered selectors virtual column?). Then we maybe we can just ditch this stuff entirely in some sunny future instead of implementing vectorized versions of decorate for the various `DimensionSpec` implementations. Or is there an advantage to `DimensionSpec` that I am missing?

Reply:
> I don't know, the more I think about this, do we really need DimensionSpec, like at all?

I don't think so, because the stuff it does could be done by virtual columns. Removing them would be a big backwards-incompatibility, but we don't necessarily have to remove them; we could just make them not-vectorizable and add notes in the docs that they are a legacy thing.

Today the only DimensionSpec that is vectorizable is DefaultDimensionSpec (the one that does nothing other than pass-through), so nothing special is needed to support it.

Reply:
I've removed the `decorate` method I added in this PR, in favor of removing this functionality from the vectorized engine. Since no `decorate`, no need for a separate "object dimension selector", I've removed that from `VectorColumnSelectorFactory` as well and reverted the signatures changes to processors that were needed to support that :+1:

If we do the same thing and ditch `decorate` for `SingleValueDimensionVectorSelector` and `MultiValueDimensionVectorSelector` there is quite a bit of simplification we can do, because we can discard the `DimensionSpec` much earlier and just work with `columnName` like the other selectors instead of going all the way down like it does now.